etcdserver/server.go

```diff
@@ -777,7 +777,55 @@ func (s *EtcdServer) triggerSnapshot(ep *etcdProgress) {
 
 // Stop stops the server gracefully, and shuts down the running goroutine.
 // Stop should be called after a Start(s), otherwise it will block forever.
+// When stopping leader, Stop transfers its leadership to one of its peers
+// before stopping the server.
 func (s *EtcdServer) Stop() {
+	if s.Cfg != nil && s.Cfg.AutoLeaderTransfer &&
+		s.cluster != nil && len(s.cluster.MemberIDs()) > 1 &&
+		uint64(s.ID()) == s.Lead() {
+		transferee := uint64(0)
+		for _, id := range s.cluster.MemberIDs() {
+			if id != s.id {
+				transferee = uint64(id)
+				break
+			}
+		}
+		if transferee != 0 {
+			plog.Printf("stopping transport at %s", s.ID())
+			s.r.transport.Stop()
+			s.PauseSending()
+
+			plog.Printf("transferring leadership from %s to %x before existing...", s.ID(), transferee)
+			ctx, cancel := context.WithCancel(context.TODO())
+			donec, stopc := make(chan struct{}), make(chan struct{})
+			go func() {
+				defer close(donec)
+				s.r.TransferLeadership(ctx, uint64(s.ID()), transferee)
+
+				interval := s.Cfg.electionTimeout()
+				for s.Lead() != transferee { // wait till transferee becomes leader
+					select {
+					case <-stopc: // timed out
+						return
+					case <-time.After(interval):
+					}
+				}
+				plog.Infof("leader transfer is done (new leader %x)", transferee)
+			}()
+
+			tm := s.Cfg.ReqTimeout()
+			plog.Infof("waiting for transferring to finish for %v...", tm)
+
```

Comment:
On second thought, that's probably not necessary; there'll only be registered waiters if the requests come in through the api layer. If there's a transfer it can be assumed the clients are accessing the cluster through the other members. It only needs to wait until the leadership changed and (maybe) the applied index is greater than the index on the re-election.